text/0000-let-else.md

```diff
@@ -0,0 +1,520 @@
+
+This RFC is a modernization of a [2015 RFC (pull request 1303)][old-rfc].
+
+A lot of this RFC's proposals come from that RFC and its ensuing discussions.
+
+The Swift programming language, which inspired Rust's if-let expression, also
+includes a [guard-let-else][swift] statement which is roughly equivalent to this
+proposal except for the choice of keywords.
+
+The `match` alternative in particular is fairly prevalent in rust code on projects which have many possible error conditions.
+
+The Try operator allows for an `ok_or_else` alternative to be used where the types are only `Option` and `Result`,
+which is considered to be idiomatic rust.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None known at time of writing due to extensive pre-discussion in Zulip:
+https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/.60let.20pattern.20.3D.20expr.20else.20.7B.20.2E.2E.2E.20.7D.60.20statements
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## if-let-chains
+
+An RFC exists for a (unimplemented at time of writing) feature called [if-let-chains][]:
+
+```rust
+if let Some(foo) = expr() && foo.is_baz() && let Ok(yay) = qux(foo) { ... }
+```
+
+While this RFC does not introduce or propose the same thing for let-else it attempts to allow it to be a future possibility for
+potential future consistency with if-let-chains.
+
+The primary obstacle is existing operator order precedence.
+Given the above example, it would likely be parsed as follows with ordinary operator precedence rules for `&&`:
+```rust
+let Some(foo) = (expr() && foo.is_baz() && let Ok(yay) = qux(foo) else { ... })
+```
+
+However, given that all existing occurrences of this behavior before this RFC are type errors anyways,
+a specific boolean-only case can be avoided and thus parsing can be changed to lave the door open to this possible extension.
+This boolean case is always equivalent to a less flexible `if` statement and as such is not useful.
+
+```rust
+let maybe = Some(2);
+let has_thing = true;
+
+// Always an error regardless, because && only operates on booleans.
+let Some(x) = maybe && has_thing else {
+    return;
+};
+```
+
+```rust
+let a = false;
+let b = false;
+
+// The RFC proposes boolean patterns with a lazy boolean operator (&& or ||)
+//  be made into a compile error, for potential future compatibility with if-let-chains.
+let true = a && b else {
+    return;
+};
+```
+
+Note also that this does not work today either, because booleans are refutable patterns:
+```
+error[E0005]: refutable pattern in local binding: `false` not covered
+ --> src/main.rs:5:9
+  |
+5 |     let true = a && b;
+  |         ^^^^ pattern `false` not covered
+  |
+  = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
+```
+
+## Fall-back assignment
+
+This RFC does not suggest that we do any of these, but notes that they would be future possibilities.
```

Comment:
Above you stated that this RFC can't be extended to include fallback assignment, but here is an idea to do so. Can you combine the two sections and remove the contradiction?

Reply:
Yes, there are a set of reasons this RFC avoids it, but it does also try to keep the door open to such a thing when possible.

Could you point to the specific text that may be more conflicting than that?

Reply:
This is the line that I thought seemed contradictory: https://github.com/rust-lang/rfcs/pull/3137/files/f3473f68686b6e0d397ceb51240ebf2cbbc2471e#diff-66dec22a01118a5d83cc226af3c8c32342a0093d60502b30b9b8f2a263e319d5R371

Reply:
That specific form of fall-back assignment is incompatible with the diverging block, yes.